lib/optimizer/src/Transformer/PreloadHeroImage.php

```diff
@@ -0,0 +1,563 @@
+    }
+
+    /**
+     * Detect a hero image candidate.
+     *
+     * The hero image here can come from one of <amp-img>, <amp-video>, <amp-iframe>, <amp-video-iframe>.
+     *
+     * @param DOMElement $element Element to detect for.
+     * @return HeroImage|null Detected hero image candidate, or null if none detected.
+     */
+    private function detectHeroImageCandidate(DOMElement $element)
+    {
+        if (
+            $element->hasAttribute(Attribute::LAYOUT)
+            && $element->getAttribute(Attribute::LAYOUT) === Layout::NODISPLAY
+        ) {
+            return null;
+        }
+
+        if ($element->tagName === Extension::IMAGE) {
+            return $this->detectHeroImageCandidateForAmpImg($element);
+        }
+
+        if ($element->tagName === EXTENSION::VIDEO) {
+            return $this->detectHeroImageCandidateForPosterImage($element);
+        }
+
+        if (Amp::isAmpIframe($element)) {
+            return $this->detectHeroImageCandidateForIframePlaceholderImage($element);
+        }
+
+        return null;
+    }
+
+    /**
+     * Detect a hero image candidate from an <amp-img> element.
+     *
+     * @param DOMElement $element Element to detect for.
+     * @return HeroImage|null Detected hero image candidate, or null if none detected.
+     */
+    private function detectHeroImageCandidateForAmpImg(DOMElement $element)
+    {
+        $src = $element->getAttribute(Attribute::SRC);
+
+        if (empty($src)) {
+            return null;
+        }
+
+        if (! Url::isValidImageSrc($src)) {
+            return null;
+        }
+
+        $width  = $element->getAttribute(Attribute::WIDTH);
+        $height = $element->getAttribute(Attribute::HEIGHT);
+        $layout = $element->getAttribute(Attribute::LAYOUT);
+
+        if (! $width && ! $height) {
+            if ($layout === Layout::FILL) {
+                list($width, $height) = $this->getDimensionsFromParent($element);
+            } else {
+                return null;
+            }
+        }
+
+        if ($this->isTinyElement($layout, $width, $height)) {
+            return null;
+        }
+
+        $srcset = $element->getAttribute(Attribute::SRCSET);
+        $media  = $element->getAttribute(Attribute::MEDIA);
+
+        return new HeroImage($src, $media, $srcset, $element);
+    }
+
+    /**
+     * Get the dimensions to use from an element's parent(s).
+     *
+     * @param DOMElement $element Element to check the parents of.
+     * @return array Array containing the width and the height.
+     */
+    private function getDimensionsFromParent(DOMElement $element)
+    {
+        while ($element->parentNode) {
+            $element = $element->parentNode;
+
+            if (! $element instanceof DOMElement) {
+                continue;
+            }
+
+            $width  = $element->getAttribute(Attribute::WIDTH);
+            $height = $element->getAttribute(Attribute::HEIGHT);
+
+            if (! $width && ! $height) {
```

Comment:
The `$width` and `$height` could be `'0'` here, and this would unexpectedly result in `continue`, no?

Also, what if the the `$width` is `'auto'` due to the parent having a `fixed-height` layout? It doesn't seem that other methods like `isTinyElement` is expecting a non-integer value.

Reply:
Yes, all valid issues. I used this code as-is from the node.js lib, but I also think that it is overly simplistic in this case.

I also think there should be a limit to how far up the dimensions are being checked for. There's no point in climbing up 20 parent nodes and then check the width and height of the `<html>` element...

Reply:
> Also, what if the the $width is 'auto' due to the parent having a fixed-height layout? It doesn't seem that other methods like isTinyElement is expecting a non-integer value.

It seems like the HTML `width` & `height` properties only support integer values that represent pixels. Are you thinking about CSS width/height here, or am I missing something?

Reply:
In AMP, floats are allowed for AMP components. Also, when non-fixed layouts, certain string identifiers can use:

* For `fixed-height`, `width` can be `auto` (or omitted).
* For `fluid` layout (which appears to only be for ads), the value of `fluid` can be used for both the `width` and `height` (as I see): https://medium.com/ampfuel/increase-ad-competition-with-multi-size-fluid-ads-in-amp-9cff90066c16

Reply:
Yes, of course, we are manipulating an `<amp-img>`, not an `<img>`. I seem to have taken a wrong turn yesterday when refactoring the `ImageDimenions`... I guessed I wanted to invent SSR for pure HTML5... 🙃